ingestion/src/legacy/raw.rs

```diff
@@ -0,0 +1,421 @@
+"#;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("kafka returned an error: {0}")]
+    Kafka(#[from] KafkaError),
+    #[error("failed to determine format of kafka message")]
+    Format,
+    #[error("failed to decode kafka message as Utf8")]
+    Utf8(#[from] std::str::Utf8Error),
+    #[error("failed to parse kldata message: {0}")]
+    Parse(#[from] ParseError),
+    #[error("postgres returned an error: {0}")]
+    Database(#[from] tokio_postgres::Error),
+    #[error(transparent)]
+    Common(#[from] common::Error),
+}
+
+type Datum = CommonDatum<ObsType>;
+type UnlabelledDatum = CommonUnlabelledDatum<ObsType>;
+
+// we have to do this with the u8 slice because some messages on the topic
+// (bufr) cannot be decoded as utf8
+fn is_kldata_message(message: &[u8]) -> Result<bool, Error> {
+    let format_raw = message
+        .split(|elem| {
+            let char = *elem as char;
+            char == '\n' || char == '/'
+        })
+        .next()
+        .ok_or(Error::Format)?;
+
+    let format = std::str::from_utf8(format_raw)?;
+
+    Ok(format.trim() == "kldata")
+}
+
+// modified version of kldata::parse_obs that returns RawDatum instead of ObsinnChunk
+fn parse_obs(
+    csv_body: Lines,
+    columns: &[ObsinnId],
+    reference_params: ParamConversions,
+    header: ObsinnHeader,
+) -> Result<Vec<UnlabelledDatum>, ParseError> {
+    let mut obs = Vec::new();
+
+    for row in csv_body {
+        let (timestamp, vals) = {
+            let mut vals = row.split(',');
+
+            let raw_timestamp = vals.next().ok_or(ParseError::EmptyRow)?;
+
+            // TODO: timestamp parsing needs to handle milliseconds and truncated timestamps?
+            let timestamp = NaiveDateTime::parse_from_str(raw_timestamp, "%Y%m%d%H%M%S")?.and_utc();
+
+            (timestamp, vals)
+        };
+
+        for (i, val) in vals.enumerate() {
+            // TODO: should we do some smart bounds-checking??
+            let col = columns[i].clone();
+
+            // rejection is acceptable here, because things we don't catch should
+            // be covered by the checked queue
+            let param_entry = reference_params.get(&col.param_code);
+
+            let (sensor, level) = col.sensor_and_level.unwrap_or((0, 0));
+
+            let val = val.trim();
+
+            // // TODO: is this right?
+            // // Ignore empty values
+            if val.is_empty() {
+                continue;
+            }
+
+            let param = match param_entry {
+                Some(entry) => Param::Id(entry.id),
+                None => Param::Code(col.param_code),
+            };
+
+            let value: ObsType = if param_entry.is_some() && !param_entry.unwrap().is_scalar {
+                ObsType::NonScalar(val.to_string())
+            } else {
+                match val.parse::<f64>() {
+                    Ok(parsed) => ObsType::Scalar(parsed),
+                    Err(_) => ObsType::NonScalar(val.to_string()),
+                }
```

Comment:
The parameter is supposed to be scalar, should the error branch actually return an error instead?
 
Edit: ah, we get to this else both if there is no `param_id` or if the parameter is a scalar, should they be treated the same?

Reply:
Yeah, IIRC when we originally wrote this parser we made plans to do a survey of all the params for this purpose, but it hasn't been a priority